### PR TITLE
Fix leaf reader and vector reader overrides

### PR DESCRIPTION
**Pull request by benwtrent**
Update leaf and vector readers. Changes in Lucene for vector search requires new `KnnCollector` interface.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/internal/FieldUsageTrackingDirectoryReader.java [hunk 1]
```diff
@@ -234,21 +234,19 @@ public ByteVectorValues getByteVectorValues(String field) throws IOException {
         }
 
         @Override
-        public TopDocs searchNearestVectors(String field, byte[] target, int k, Bits acceptDocs, int visitedLimit) throws IOException {
-            TopDocs topDocs = super.searchNearestVectors(field, target, k, acceptDocs, visitedLimit);
-            if (topDocs != null) {
+        public void searchNearestVectors(String field, byte[] target, KnnCollector collector, Bits acceptDocs) throws IOException {
+            super.searchNearestVectors(field, target, collector, acceptDocs);
+            if (collector.visitedCount() > 0) {
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/search/internal/FieldUsageTrackingDirectoryReader.java:239]: This is the biggest change here. If the collector doesn't visit any documents, that means there weren't matching the `acceptDocs` (match_none), there aren't any docs with that field, or the field just isn't there. 

I am not sure the usage of `FieldUsageTrackingDirectoryReader` and how we need to handle this :/

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/search/internal/FieldUsageTrackingDirectoryReader.java:239]: This changes the behavior slightly, before even if we didn't match anything, we would track the field usage. 

Also, previously, if we used `byte[]` to query a `float[]` field, it returned `null`. But, honestly, that seems like a bug to me :/.

Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/search/internal/FieldUsageTrackingDirectoryReader.java:239]: If we want to keep the behaviour the same I think we can call `super.searchNearestVectors` with a wrapped `KnnCollector` that calls `notifier.onKnnVectorsUsed(field)` in its `collect()` method

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/search/internal/FieldUsageTrackingDirectoryReader.java:239]: Boom! 100% you are correct. I will do this.

**Comment by benwtrent** [comment]
This: https://github.com/apache/lucene/pull/12481

Seems to have broken profiling. Digging into all the profiling code, it seems like `advance` is no longer being called on our scorer iterators after this change.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/370_profile.yml [hunk 2]
```diff
@@ -220,8 +220,6 @@ dfs knn vector profiling:
   - gt: { profile.shards.0.dfs.knn.0.query.0.breakdown.score: 0 }
   - match: { profile.shards.0.dfs.knn.0.query.0.breakdown.compute_max_score_count: 0 }
   - match: { profile.shards.0.dfs.knn.0.query.0.breakdown.compute_max_score: 0 }
-  - gt: { profile.shards.0.dfs.knn.0.query.0.breakdown.advance_count: 0 }
-  - gt: { profile.shards.0.dfs.knn.0.query.0.breakdown.advance: 0 }
```
Comment (benwtrent) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/370_profile.yml:224]: This: https://github.com/apache/lucene/pull/12481

Seems to have broken profiling. Digging into all the profiling code, it seems like `advance` is no longer being called on our scorer iterators after this change.

Looks like it calls `nextDoc`, and doesn't forward the call to `advance(docId + 1)` any longer all the time. Seems ok to me.

**Pull request merged**